hazelcast/src/main/java/com/hazelcast/concurrent/flakeidgen/FlakeIdGeneratorProxy.java

```diff
@@ -0,0 +1,198 @@
+            @Override
+            public IdBatch apply(Integer batchSize) {
+                return newIdBatch(batchSize);
+            }
+        });
+    }
+
+    @Override
+    public long newId() {
+        return batcher.newId();
+    }
+
+    @Override
+    public IdBatch newIdBatch(int batchSize) {
+        while (true) {
+            NewIdBatchOperation op = new NewIdBatchOperation(name, batchSize);
+            Member member = getRandomMember();
+            InternalCompletableFuture<Long> future = getNodeEngine().getOperationService()
+                                                                    .invokeOnTarget(getServiceName(), op, member.getAddress());
+            try {
+                long base = future.join();
+                return new IdBatch(base, 1 << BITS_NODE_ID, batchSize);
+            } catch (FlakeIdNodeIdOutOfRangeException e) {
+                outOfRangeMembers.add(member.getUuid());
+                randomMember = null;
+            }
+        }
+    }
+
+    long newIdBaseLocal(int batchSize) {
+        return newIdBaseLocal(Clock.currentTimeMillis(), batchSize);
+    }
+
+    /**
+     * The layout of the ID is as follows (starting from most significant bits):<ul>
+     *     <li>42 bits timestamp
+     *     <li>6 bits sequence
+     *     <li>16 bits node ID
+     * </ul>
+     *
+     * This order is important: timestamp must be first to keep IDs ordered. Sequence must be second for
+     * implementation reasons (it's included in {@link #generatedValue}). Node is just an appendix to make
+     * IDs unique.
+     *
+     * @param now Current time (currentTimeMillis() normally or other value in tests)
+     * @return First ID for the batch.
+     */
+    long newIdBaseLocal(long now, int batchSize) {
+        checkPositive(batchSize, "batchSize");
+        if (nodeId < 0) {
+            logger.info("Throwing FlakeIdNodeIdOutOfRangeException");
+            throw new FlakeIdNodeIdOutOfRangeException("NodeID overflow, this member cannot generate IDs");
+        }
+        logger.info("Generating " + batchSize + " IDs");
+        now -= EPOCH_START;
+        assert now >= -(1L << BITS_TIMESTAMP - 1) && now < (1L << BITS_TIMESTAMP - 1)  : "Current time out of allowed range";
+        now <<= BITS_SEQUENCE;
+        long oldGeneratedValue;
+        long base;
+        do {
+            oldGeneratedValue = generatedValue.get();
+            base = Math.max(now, oldGeneratedValue);
+        } while (!generatedValue.compareAndSet(oldGeneratedValue, base + batchSize));
+        return base << BITS_NODE_ID | nodeId;
+    }
+
+    private Member getRandomMember() {
+        Member member = randomMember;
+        if (member == null) {
+            // use local member if possible. Local member is always the first one inserted to outOfRangeMembers
+            if (outOfRangeMembers.isEmpty()) {
+                member = getNodeEngine().getLocalMember();
```

Comment:
Can you add a new boolean that will say if the local member has overflown or not (or do the check here) and use the local member if not. Seems that as soon as one member has overflown (even if its not the local member), all requests will be forwarded to a random (possibly remote) member.